Update Node rx type enum to be self-incrementing

Thalley opened the pull request:
This PR removes the #ifdefs from the enum and removes all the explicitly set values. This is to make the enum self-incrementing, while still having the same values regardless of the configuration. 

Making the enum self-incrementing will reduce manual change to be done when a new entry is added, and will also remove any risk of duplicate values.

Thalley committed:
Bluetooth: controller: Remove #ifdef for node_rx_type enum

Removed the #ifdef's for the node_rx_type enum. Since the
savings, they just make the enum harder to read.

Furthermore it will also ensure that all the enum values
will have the same value once the enum entries are no
longer manually defined.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Controller: Change node_rx_type to auto incrementing enum

Instead of having explicit values for each entry in the enum that
has to be updated for each new entry (thus requiring manual work
which can also be error-prone), the enum will now be
self-incrementing, to make it easier and safer to add new
entries.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

thoh-ot commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -155,94 +155,51 @@ enum node_rx_type {
+ * used to explicitly initialize members other than the first, unless all items
+ * are explicitly initialized.
+ */
+#define NODE_RX_TYPE_USER_END = NODE_RX_TYPE_USER_START +
```
The '=' should be removed for this to even compile.

I understand what you are trying to do, but it gives warnings when compiling with CONFIG_BT_CTLR_USER_EXT=y as NODE_RX_TYPE_USER_END is now no longer a value in the enumerated type.

warning: case value ‘X’ not in enumerated type ‘enum node_rx_type’ [-Wswitch]                                         
case NODE_RX_TYPE_USER_START ... NODE_RX_TYPE_USER_END:

Not sure how i like those warnings, could be a problem if you were dedicated to the -Werror style of life...

thoh-ot replied:
I have an idea on how to solves this: we need a pre-processor enum value generator thingy...

As `CONFIG_BT_CTLR_USER_EVT_RANGE` has the range 0 to 10 the following construction can genrate the padding need between `NODE_RX_TYPE_USER_START` and `NODE_RX_TYPE_USER_END`

It's not entirely the same as `NODE_RX_TYPE_USER_START` != `NODE_RX_TYPE_USER_DUMMY_1` and `NODE_RX_TYPE_USER_END` != `NODE_RX_TYPE_USER_DUMMY_5` if `CONFIG_BT_CTLR_USER_EVT_RANGE=5`, but that should be ok.

```
#define NODE_RX_TYPE_USER_GENERATOR_0(X)
#define NODE_RX_TYPE_USER_GENERATOR_1(X)	 X ## 1,
#define NODE_RX_TYPE_USER_GENERATOR_2(X)	 NODE_RX_TYPE_USER_GENERATOR_1(X) X ## 2,
#define NODE_RX_TYPE_USER_GENERATOR_3(X)	 NODE_RX_TYPE_USER_GENERATOR_2(X) X ## 3,
#define NODE_RX_TYPE_USER_GENERATOR_4(X)	 NODE_RX_TYPE_USER_GENERATOR_3(X) X ## 4,
#define NODE_RX_TYPE_USER_GENERATOR_5(X)	 NODE_RX_TYPE_USER_GENERATOR_4(X) X ## 5,
#define NODE_RX_TYPE_USER_GENERATOR_6(X)	 NODE_RX_TYPE_USER_GENERATOR_5(X) X ## 6,
#define NODE_RX_TYPE_USER_GENERATOR_7(X)	 NODE_RX_TYPE_USER_GENERATOR_6(X) X ## 7,
#define NODE_RX_TYPE_USER_GENERATOR_8(X)	 NODE_RX_TYPE_USER_GENERATOR_7(X) X ## 8,
#define NODE_RX_TYPE_USER_GENERATOR_9(X)	 NODE_RX_TYPE_USER_GENERATOR_8(X) X ## 9,
#define NODE_RX_TYPE_USER_GENERATOR_10(X)	 NODE_RX_TYPE_USER_GENERATOR_9(X) X ## 10,
#define NODE_RX_TYPE_USER_GENERATOR_EXPAND(N, X) NODE_RX_TYPE_USER_GENERATOR_##N(X)
#define NODE_RX_TYPE_USER_GENERATOR(N, X)	 NODE_RX_TYPE_USER_GENERATOR_EXPAND(N, X)
```

```
enum node_rx_type {
...
	NODE_RX_TYPE_USER_START,
	NODE_RX_TYPE_USER_GENERATOR(CONFIG_BT_CTLR_USER_EVT_RANGE, NODE_RX_TYPE_USER_DUMMY_)
	NODE_RX_TYPE_USER_END,
};
```

Thalley replied:
Good catch! I guess CI doesn't have a build config with `CONFIG_BT_CTLR_USER_EXT=y` since it didn't catch this... 

I understand your worry, and I would really like to fix it. However for MISRA we cannot have it at the end of the enumeration either. Not sure how big a deal conforming to MISRA is. 

When I build with `west`, I actually get an error rather than a warning.

thoh-ot replied:
There actually is a "compile test" in tests/bluetooth/ctrl_user_ext that uses `CONFIG_BT_CTLR_USER_EXT=y`, but i guess CI doesn't compile that one.

jukkar replied:
There is `UTIL_LISTIFY()` macro that does something like your generator. Can't you just do it like this (not tested)

```
#define DEFINE_NODE_RX_USER_TYPE(i, _) NODE_RX_TYPE##i,
enum node_rx_type {
   NODE_RX_TYPE_USER_START,
   UTIL_LISTIFY(CONFIG_BT_CTLR_USER_EVT_RANGE, DEFINE_NODE_RX_USER_TYPE, _)
   NODE_RX_TYPE_USER_END,
};
```

Thalley replied:
Was about to say the same as @jukkar. I'll try to see if I can cook up a version using that, as that would be preferable in any way I suppose :)

Thalley commented:
Updated the PR as it gave errors/warnings with `NODE_RX_TYPE_USER_END` not being in the enum. This however, means that we no longer comply with: 

> MISRA C:2004, 9.3 - In an enumerator list, the "=" construct shall not be used to explicitly initialize members other than the first, unless all items are explicitly initialized.

thoh-ot commented:
Did you look at the pre-processor thing i suggested or i that too ugly?

Thalley commented:
> Updated the PR as it gave errors/warnings with `NODE_RX_TYPE_USER_END` not being in the enum. This however, means that we no longer comply with:
> 
> > MISRA C:2004, 9.3 - In an enumerator list, the "=" construct shall not be used to explicitly initialize members other than the first, unless all items are explicitly initialized.

Updated the PR with the suggestion from Jukkar. Regarding the MISRA; there's plenty of other enums that doesn't comply with that, so would not have a big issue if we had it simply had it as before.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -151,96 +152,49 @@ struct lll_event {
-				CONFIG_BT_CTLR_USER_EVT_RANGE,
+	/* No entries shall be added after the NODE_RX_TYPE_USER_START/END */
+	NODE_RX_TYPE_USER_START,
+	UTIL_LISTIFY(CONFIG_BT_CTLR_USER_EVT_RANGE, DEFINE_NODE_RX_USER_TYPE, _)
```
I think there is an off-by-one error here. If `CONFIG_BT_CTLR_USER_EVT_RANGE`=0, `NODE_RX_TYPE_USER_END` should get the same value as `NODE_RX_TYPE_USER_START` but it will get the value+1 with this change.

thoh-ot replied:
I don't think that will be a problem, `NODE_RX_TYPE_USER_END` is only used in the case `NODE_RX_TYPE_USER_START ... NODE_RX_TYPE_USER_END`, so it catches an extra value, so you'll get CONFIG_BT_CTLR_USER_EVT_RANGE+1 user defined node rx types.

wopu-ot replied:
I guess changing the cases to `NODE_RX_TYPE_USER_START ... NODE_RX_TYPE_USER_END-1` would avoid the off-by-one then.

Thalley replied:
@wopu-ot We can also guard this part with `CONFIG_BT_CTLR_USER_EVT_RANGE > 0` and then change the list to 
```
	UTIL_LISTIFY(CONFIG_BT_CTLR_USER_EVT_RANGE, DEFINE_NODE_RX_USER_TYPE, _)
```

Would you prefer that?

thoh-ot replied:
UTIL_LISTIFY doesn't work with numeric expressions.

Thalley replied:
I see. I guess a hidden Kconfig that is 1 less than `CONFIG_BT_CTLR_USER_EVT_RANGE` would "fix" that, but then we're just walking towards making things unnecessary complicated IMO. 

Or do as suggested with `NODE_RX_TYPE_USER_START ... NODE_RX_TYPE_USER_END - 1`

wopu-ot replied:
I prefer the latter. Counting/iterating from 0 to N-1 is a perfectly normal and non-hacky way of doing things.

Thalley replied:
Updated the PR. 

Speaking of off-by-one, I wonder if that is still the case (and always was), since if `CONFIG_BT_CTLR_USER_EVT_RANGE=0` we still have 1 case in the switches that would match (if `type == NODE_RX_TYPE_USER_START`). 

Should we guard 
```
case NODE_RX_TYPE_USER_START ... NODE_RX_TYPE_USER_END - 1:
```
with 
```
#if CONFIG_BT_CTLR_USER_EVT_RANGE > 0
```
?

thoh-ot replied:
That would be a good idea

Thalley committed:
Bluetooth: controller: guard NODE_RX_TYPE_USER with range value

Instead of guarding the NODE_RX_TYPE_USER cases with
CONFIG_BT_CTLR_USER_EXT we guard them with
CONFIG_BT_CTLR_USER_EVT_RANGE > 0 as that is the actual value that
enables/disables the NODE_RX_TYPE_USER values.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>
